### ISSUE TITLE
Unable to instantiate Basecamp3 class with arguments.

### ISSUE THREAD

**Issue by slash5k1** [phistrom/basecampy3#3]
Looking at the bc3_api.py it would appear I should be able to pass in arguments to create a Basecamp3 object without the need for creating a configuration file first.

ie:

bc3 = Basecamp3(client_id='blahblahblah', 
client_secret='blahblahblah', 
redirect_uri='https://blah.com/auth',
refresh_token='blahblahblah'
)

When I try this I get the following error:

```
Traceback (most recent call last):
  File "/Users/canders/Code/reports/basereport.py", line 109, in <module>
refresh_token='blahblahblah',
  File "/Users/canders/Code/python3-testing/lib/python3.6/site-packages/basecampy3/bc3_api.py", line 110, in __init__
    self._authorize()
  File "/Users/canders/Code/python3-testing/lib/python3.6/site-packages/basecampy3/bc3_api.py", line 170, in _authorize
    self._get_access_token()
  File "/Users/canders/Code/python3-testing/lib/python3.6/site-packages/basecampy3/bc3_api.py", line 181, in _get_access_token
    self._save_token_json(token_json)
  File "/Users/canders/Code/python3-testing/lib/python3.6/site-packages/basecampy3/bc3_api.py", line 236, in _save_token_json
    self._conf.save()
  File "/Users/canders/Code/python3-testing/lib/python3.6/site-packages/basecampy3/config.py", line 78, in save
    os.makedirs(os.path.dirname(filepath), mode=0o770)
  File "/Users/canders/Code/python3-testing/bin/../lib/python3.6/posixpath.py", line 154, in dirname
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

Looks like you have logic that assumes we need to save these details which are then read when creating the API calls to basecamp?

Looking at:
```
bc3_api.py", line 236, in _save_token_json
config.py", line 78, in save
```

the problem looks to be that you are not passing in the filepath argument hence the NoneType... I assume this was an oversight as you have in the BasecampConfig constructor self.filepath defined.

Can I suggest that you update the constructor for the Basecamp3 class to include filepath to be passed in to _save_token_json OR could we look at making use of io.StringIO() so that we dont have to worry about saving a file when arguments are being passed in?

I like the second option so I could run this code in a container such as google's app engine :)

**Comment by slash5k1** [comment]
Here is my quick prototype for using io.StringIO to replace the need to save an actual file somewhere:

Updated config.py with:

```
import io

class BasecampConfig(object):

    memoryfile = io.StringIO()

    DEFAULT_CONFIG_FILE_LOCATIONS = [
        # "basecamp.conf",  # current directory
        # os.path.expanduser("~/.conf/basecamp.conf"),  # user profile directory/.conf/basecamp.conf
        # "/etc/basecamp.conf",
        memoryfile
    ]
```

```
    def save(self, filepath=None):
        if filepath is None:
            filepath = self.filepath
        # try:
        #     os.makedirs(os.path.dirname(filepath), mode=0o770)
        # except OSError:
        #     pass  # folder probably already exists

        config = ConfigParser()
        config.add_section('BASECAMP')
        attrs = [k for k in self.__dict__.keys() if k != 'filepath']
        for key in attrs:
            try:
                value = getattr(self, key)
                if value is None:
                    continue
                if key == "_access_expires":
                    key = "access_expires"
                    try:
                        value = value.timestamp()
                    except:
                        continue  # if timestamp is at min value, this will fail, that's ok

                config.set('BASECAMP', key, six.text_type(value))
                setattr(self, key, value)
            except (NoSectionError, NoOptionError):
                pass

        with BasecampConfig.memoryfile as fileout:
            config.write(fileout)
        # with open(filepath, "w") as fileout:
        #     config.write(fileout)
        self.filepath = filepath
```

updated bc3_api.py with:
```
    def _save_token_json(self, token_json):
        if 'access_token' in token_json:
            self._conf.access_token = token_json['access_token']
        if 'refresh_token' in token_json:
            self._conf.refresh_token = token_json['refresh_token']
        if 'expires_in' in token_json:
            self._conf.access_expires = token_json['expires_in']
        self._conf.save(filepath=BasecampConfig.memoryfile) 
```

After making these changes I was able to construct the Basecamp3 object with arguments without the need for a configuration file or saving a file.

**Comment by phistrom** [comment]
I appreciate you taking the time to file this issue. After an obscenely long wait, I'm happy to report this should be resolved in 0.3.0.

BasecampConfig is now an abstract class that you can subclass to make your own configuration object. BasecampMemoryConfig and BasecampFileConfig are included. The memory config is used when you pass parameters directly into the Basecamp3 object.
